extensions/playerlist/common/client.mjs

```diff
@@ -28,49 +28,89 @@ class Client {
    */
   constructor(helper) {
     this.helper = helper;
+    this.template = document.getElementById('playerlist-player').cloneNode(true);
+    this.playerList = {};    
   }
 
   async gameStart() {
     this.tab = new PlayerlistTab(this.helper.mainUI);
+    this.initPlayerList();
+    this.helper.gameState.registerOnPlayerUpdate(msg => this.syncPlayerList(msg));
+    this.updatePlayerList();
+  }
+
+  /**
+   * init the playlist from the server
+   * @todo Get user role.
+   * @todo Get user picture.
+   */
+  initPlayerList() {
+    this.helper.gameState.getPlayers().forEach(player => {
+      this.playerList[player.playerID] = {
+        playerID : player.playerID, 
+        displayName : player.displayName, 
+        displayChar : player.displayChar
+      }
+    });
+  }
+
+  /**
+   * Get the PlayerSyncMessage from server and update the playerList object.
+   * @param {object} msg - The PlayerSyncMessage from server.
+   * @todo Get user role.
+   * @todo Get user picture.
+   */
+  syncPlayerList(msg) {
+    if (msg.removed == true) {
+      delete this.playerList[msg.playerID];
+      return;
+    }
 
-    this.showPlayerList();
+    if (Object.keys(this.playerList).includes(msg.playerID) == false) {
+      this.playerList[msg.playerID] = {
+        playerID : msg.playerID, 
+        displayName : msg.displayName, 
+        displayChar : msg.displayChar
+      }
+    } else {
+      return;
+    }
   }
 
   /**
-   * Show player list on left top.
-   * @showPlayerList
+   * refresh the playerlist dom.
+   * @updatePlayerList
+   * @todo Get user role.
+   * @todo Get user picture.
    */
-  showPlayerList() {
-    const template = document.getElementById('playerlist-player').cloneNode(true);
-    // TODO(zeze-zeze): Optimize the way to update playerlist
+  updatePlayerList() {
     setInterval(() => {
       document.getElementById('playerlist').innerHTML = '';
-      this.helper.gameState.getPlayers().forEach(playerID => {
-        if (typeof playerID.displayName !== 'string' || !playerID.displayName.includes(document.getElementById("searchPlayer").value)) {
-          return;
+      for (const playerID of Object.keys(this.playerList)) {
+        if (typeof this.playerList[playerID].displayName !== 'string' || !this.playerList[playerID].displayName.includes(document.getElementById("searchPlayer").value)) {
```

Comment:
If the type of `this.playerList` is `Object`:
```javascript
for (const [playerID, player] of Object.entries(this.playerList)) {
  if (typeof player.displayName !== 'string' || player.displayName.includes(this.searchPlayerDOM.value)) {
```

If the type of `this.playerList` is `Map`:
```javascript
for (const [playerID, player] of this.playerList.entries()) {
  if (typeof player.displayName !== 'string' || player.displayName.includes(this.searchPlayerDOM.value)) {
```